File: Problem2_ReplaceWords.py
'''
====== Submission Details =======
Student Name: Pavan Kumar K. N.
S30 SlackID : RN32MAY2021
=================================
'''

# 648. Replace Words

# https://leetcode.com/problems/replace-words/

#-----------------
# Time Complexity: 
#-----------------
# O(NM) - N is the length of the sentence. M is the max key size

#------------------
# Space Complexity: 
#------------------
# O(N): Size of the trie

#-----------------------
# Leet Code Performance: 
#-----------------------
# Ran Successfully?: Yes

class Solution:
    
    class Trie:
    
        class TrieNode:
            def __init__(self):
                self.isEnd = False
                self.children = [None for i in range(26)]

        def __init__(self):
            """
            Initialize your data structure here.
            """


            self.root = self.TrieNode()
            return

        def insert(self, word: str) -> None:
            """
            Inserts a word into the trie.
            """
            curr = self.root
            for i in range(len(word)):
                c = word[i]
                if curr.children[ord(c) - ord('a')] is None:
                    curr.children[ord(c)-ord('a')] = self.TrieNode() 
                curr = curr.children[ord(c)-ord('a')]

            curr.isEnd = True
    
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        self.root = self.Trie()
        for word in dictionary:
            self.root.insert(word)
        
        split_sent = sentence.split()
        result = []
        for word in split_sent:
            replacement = ""
            curr = self.root.root
            for i in range(len(word)):
                c = word[i]
                if curr.children[ord(c)-ord('a')] is None or curr.isEnd:
                    
                    break
                curr = curr.children[ord(c) - ord('a')]
                replacement += c
                
            if curr.isEnd:
                #Found smallest prefix in the sentence
                result.append(replacement)

            else:
                #Didn't find prefix of word in dictionary
                result.append(word)
        
        return " ".join(result)

File: Problem2_ReplaceWords.java
// Time Complexity : O(NK + L) [N: words, K: characters]
// Space Complexity : O(NK +L)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no


class Solution {
    
     class TrieNode {
         boolean isEnd;
        TrieNode[] children;
        
        public TrieNode()
        {
            children=new TrieNode[26];
        }
        
    }
    
    TrieNode root;
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        
        TrieNode curr=root;
        
        for(int i=0;i<word.length();i++)
        {
            char c=word.charAt(i);
            
            if(curr.children[c-'a']==null)
                curr.children[c-'a']=new TrieNode();
            
            curr=curr.children[c-'a'];
        }
        curr.isEnd=true;
    }
    
    public String replaceWords(List<String> dictionary, String sentence) {
        
        if(dictionary==null || dictionary.size()==0)
            return sentence;
        
        root=new TrieNode();
        
        for(String word:dictionary)
        {
            insert(word);
        }
        
        String[] splitArray=sentence.split("\\s");
        
        StringBuilder result=new StringBuilder();
        
        
        
        for(int j=0;j<splitArray.length;j++)
        {
            if(j>0)
                result.append(' ');
         
            TrieNode curr=root;
            String word=splitArray[j];
            StringBuilder replacement=new StringBuilder();
            for(int k=0;k<word.length();k++)
            {
                      char c=word.charAt(k);
                
                if(curr.children[c-'a']==null || curr.isEnd)
                    break;
                curr=curr.children[c-'a'];
                replacement.append(c);
            }
            if(curr.isEnd)
            {
                // reached end of word did find prefix
                result.append(replacement.toString());
            }
            else
            {
                //if we dont find
                result.append(word);
                
            }
        }
        
        
        return result.toString();
    }
}